tls/s2n_post_handshake.c

```diff
@@ -13,66 +13,172 @@
-        POSIX_GUARD(s2n_stuffer_skip_write(&post_handshake_stuffer, message_length));
-
-        switch (post_handshake_id)
-        {
-            case TLS_KEY_UPDATE:
-                POSIX_GUARD(s2n_key_update_recv(conn, &post_handshake_stuffer));
-                break;
-            case TLS_SERVER_NEW_SESSION_TICKET:
-                POSIX_GUARD_RESULT(s2n_tls13_server_nst_recv(conn, &post_handshake_stuffer));
-                break;
-            case TLS_HELLO_REQUEST:
-                POSIX_GUARD_RESULT(s2n_client_hello_request_recv(conn));
-                break;
-            case TLS_CLIENT_HELLO:
-            case TLS_SERVER_HELLO:
-            case TLS_END_OF_EARLY_DATA:
-            case TLS_ENCRYPTED_EXTENSIONS:
-            case TLS_CERTIFICATE:
-            case TLS_SERVER_KEY:
-            case TLS_CERT_REQ:
-            case TLS_SERVER_HELLO_DONE:
-            case TLS_CERT_VERIFY:
-            case TLS_CLIENT_KEY:
-            case TLS_FINISHED:
-            case TLS_SERVER_CERT_STATUS:
-                /* All other known handshake messages should be rejected */
-                POSIX_BAIL(S2N_ERR_BAD_MESSAGE);
-                break;
-            default:
-                /* Ignore all other messages */
-                break;
+    switch (message_type) {
+        case TLS_KEY_UPDATE:
+            RESULT_GUARD_POSIX(s2n_key_update_recv(conn, in));
+            break;
+        case TLS_SERVER_NEW_SESSION_TICKET:
+            RESULT_GUARD(s2n_tls13_server_nst_recv(conn, in));
+            break;
+        case TLS_HELLO_REQUEST:
+            RESULT_GUARD(s2n_client_hello_request_recv(conn));
+            break;
+        default:
+            /* Ignore unknown messages */
+            RESULT_ENSURE(!s2n_post_handshake_is_known(message_type), S2N_ERR_BAD_MESSAGE);
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_post_handshake_message_recv(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+
+    struct s2n_stuffer *in = &conn->in;
+    struct s2n_stuffer *message = &conn->post_handshake.in;
+    uint8_t message_type = 0;
+    uint32_t message_len = 0;
+
+    /* We always start reading from the beginning of the message */
+    RESULT_GUARD_POSIX(s2n_stuffer_reread(message));
+
+    /* If no space for the fragment exists, start with the minimum, static space */
+    if (s2n_stuffer_is_freed(message)) {
+        struct s2n_blob b = { 0 };
+        RESULT_GUARD_POSIX(s2n_blob_init(&b, conn->post_handshake.in_bytes,
+                sizeof(conn->post_handshake.in_bytes)));
+        RESULT_GUARD_POSIX(s2n_stuffer_init(message, &b));
+    }
+
+    /* Try to read the header */
+    if (s2n_stuffer_data_available(message) < TLS_HANDSHAKE_HEADER_LENGTH) {
+        uint32_t remaining = TLS_HANDSHAKE_HEADER_LENGTH - s2n_stuffer_data_available(message);
+        uint32_t to_read = MIN(remaining, s2n_stuffer_data_available(in));
```

Comment:
I tried to answer this with the rewritten comments. We're not worried that there's enough data in "in", we're accounting for any data already in "message".

I didn't use some of your suggested comments verbatim because I'm worried they were a little misleading in this way. Like, we don't read the whole header, we read the remaining header. And we might not allocate memory even for a large message, because we might have already allocated the memory.